tests/channel_test.c

```diff
@@ -710,3 +714,153 @@ static int s_test_channel_cancels_pending_tasks(struct aws_allocator *allocator,
+    aws_default_dns_resolve(allocator, amazon_com, &addresses, NULL);
+    ASSERT_INT_EQUALS(1, aws_array_list_length(&addresses));
+    aws_array_list_get_at_ptr(&addresses, (void*)&resolved_address, 0);
+    addr1_ipv4 = aws_string_new_from_string(allocator, resolved_address->address);
+    aws_string_destroy(amazon_com);
+
+    /* create a resolver with 2 addresses: 1 IPv4 which will always succeed, and 1 IPv6 which will always timeout */
+    struct mock_dns_resolver mock_dns_resolver;
+    ASSERT_SUCCESS(mock_dns_resolver_init(&mock_dns_resolver, 2, allocator));
+
+    struct aws_host_resolution_config mock_resolver_config = {
+        .max_ttl = 1,
+        .impl = mock_dns_resolve,
+        .impl_data = &mock_dns_resolver,
+    };
+
+    struct aws_host_address host_address_1_ipv4 = {
+        .address = addr1_ipv4,
+        .allocator = allocator,
+        .expiry = 0,
+        /* connections should always succeed, if not, things are worse than this unit test failing */
+        .host = aws_string_new_from_c_str(allocator, "www.amazon.com"),
+        .connection_failure_count = 0,
+        .record_type = AWS_ADDRESS_RECORD_TYPE_A,
+        .use_count = 0,
+        .weight = 0,
+    };
+
+    struct aws_host_address host_address_1_ipv6 = {
+        .address = addr2_ipv6,
+        .allocator = allocator,
+        .expiry = 0,
+        /* same black-holed host from the timeout test, connections are a guaranteed timeout */
+        .host = aws_string_new_from_c_str(allocator, "ec2-54-158-231-48.compute-1.amazonaws.com"),
+        .connection_failure_count = 0,
+        .record_type = AWS_ADDRESS_RECORD_TYPE_AAAA,
+        .use_count = 0,
+        .weight = 0,
+    };
+
+    struct aws_array_list address_list;
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(&address_list, allocator, 2, sizeof(struct aws_host_address)));
+    ASSERT_SUCCESS(aws_array_list_push_back(&address_list, &host_address_1_ipv6));
+    ASSERT_SUCCESS(aws_array_list_push_back(&address_list, &host_address_1_ipv4));
+    ASSERT_SUCCESS(mock_dns_resolver_append_address_list(&mock_dns_resolver, &address_list));
+
+    struct aws_client_bootstrap bootstrap;
+    aws_client_bootstrap_init(&bootstrap, allocator, &event_loop_group, NULL, &mock_resolver_config);
+
+    struct aws_socket_options options;
+    AWS_ZERO_STRUCT(options);
+    options.connect_timeout_ms = 3000;
+    options.type = AWS_SOCKET_STREAM;
+
+    struct channel_connect_test_args callback_data = {
+        .mutex = &mutex,
+        .cv = AWS_CONDITION_VARIABLE_INIT,
+        .error_code = 0,
+        .channel = NULL,
+        .setup = false,
+        .shutdown = false,
+    };
+
+    aws_client_bootstrap_new_socket_channel(
+        &bootstrap,
+        "www.amazon.com",
+        80,
+        &options,
+        s_test_channel_connect_some_hosts_timeout_setup,
+        s_test_channel_connect_some_hosts_timeout_shutdown,
+        &callback_data);
+
+    aws_condition_variable_wait(&callback_data.cv, &mutex);
```

Comment:
I don't see where you lock the mutex before waiting on the CVAR. Also, you need to use a predicate to avoid spurious wakeups.

Reply:
fixed